util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,249 @@
+) -> Path:
+    """
+    This function returns the local filepath for a resource
+    :param resource_url: URL field of the resource
+    :param base_output_path: Base output directory absolute path
+    :return: Local filepath for the resource
+    """
+    filename = f"{resource_id}-{resource_version}"
+
+    filepath = Path(base_output_path).joinpath(filename)
+
+    return filepath
+
+
+def download_resources(resources: List[Dict], output: TextIO):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    for resource in resources:
+        if "url" in resource.keys():
+            url = resource["url"]
+            filepath = get_resource_local_filepath(
+                resource["id"], resource["resource_version"], path
+            )
+
+            if not filepath.exists() or (  # If the file does not exist
+                "url-md5sum"
+                in resource.keys()  # If it exist but the md5sum is wrong.
+                and hashlib.md5(filepath.read_bytes()).hexdigest()
+                != resource["url-md5sum"]
+            ):
+                print(f"Downloading {url} to {filepath}")
+                request.urlretrieve(url, filepath)
+            resource["url"] = filepath.absolute().as_uri()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Get resources from the database"
+    )
+
+    parser.add_argument(
+        "--config-file-path",
+        type=str,
+        default=Path(__file__)
+        .resolve()
+        .parent.joinpath("gem5_default_config.py"),
+        help="Filepath to the gem5 config file",
+    )
+
+    parser.add_argument(
+        "--output_dir",
+        type=str,
+        default=Path.cwd(),
+        help="Output directory path, default is the cwd."
+        "The resources.json and all resources will be saved in this directory",
+    )
+    args = parser.parse_args()
+
+    output_path = Path(args.output_dir)
+
+    # Get the gem5 config file from the  file
+
+    with open(args.config_file_path) as f:
+        gem5_config_file_contents = f.read()
+    gem5_config_str = gem5_config_file_contents.split("=")[-1]
+    gem5_config_json = eval(gem5_config_str)
+    gem5_config_gem5_resources = gem5_config_json["sources"]["gem5-resources"]
+
+    # Parse the gem5 config
+    db_url = gem5_config_gem5_resources["url"]
+    data_source = gem5_config_gem5_resources["dataSource"]
+    collection_name = gem5_config_gem5_resources["collection"]
+    db_name = gem5_config_gem5_resources["database"]
+    auth_url = gem5_config_gem5_resources["authUrl"]
+    api_key = gem5_config_gem5_resources["apiKey"]
+
+    if not output_path.exists():
+        output_path = output_path.mkdir()
```

Comment:
This is a bit dangerous. This will only create one directory the last directory in the path, and none of that directory's parents. So, if the value of `output_path` is "x/y/z" and only an empty directory "x" exists, `output_path.exits()` returns `False` but `output_oath.mkdir()` will raise an exception because "z" cannot be created without "y" first being created.

Then solution to this is to use `os.makedirs(<path>)` instead. `os.makedirs` recursively creates the directories.

Reply:
I can change this to `output_path.mkdir(parent=True, exist_ok= True)` and this will create all the parents.

Quoting from the pathlib documentation: "If parents is true, any missing parents of this path are created as needed; they are created with the default permissions without taking mode into account (mimicking the POSIX mkdir -p command)."

link: https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir